### PR TITLE
Refactor import generation to be much more robust

### LINKED ISSUE #89: Use non-vendor import path when creating a mock for an interface in vendor/
Currently mockery uses the full vendor path, even though it would almost never make sense to directly import a path in a vendor directory. This should be fixed.

### PR DESCRIPTION
**Pull request by IvanMalison**
Fixes:  #89 #47 #94 #77 #96

### CODE REVIEW

File: mockery/generator.go [hunk 1]
```diff
@@ -11,94 +12,186 @@ import (
 	"strings"
 	"unicode"
 
-	"golang.org/x/tools/imports"
+	"code.uber.internal/go-common.git/x/log"
```
Comment (IvanMalison) [inline mockery/generator.go]: Oops, did not mean to add this. Will fix.

**Comment by IvanMalison** [comment]
@evanphx any chance you can take a look at this. This diff is getting pretty massive, but I'm pretty sure that it makes mockery much more reliable.

**Comment by IvanMalison** [comment]
@vektra @evanphx please merge!

**Comment by kevinburke** [comment]
hey Ivan - I tried pulling in your changes here: https://github.com/kevinburke/mockery/tree/ivan-vendor

and got the following failing tests:

```
$ go test ./...
?   	github.com/kevinburke/mockery	[no test files]
ok  	github.com/kevinburke/mockery/cmd/mockery	0.015s
--- FAIL: TestGeneratorPrologue (1.04s)
        Error Trace:    generator_test.go:64
			generator_test.go:194
	Error:		Not equal: "package mocks\n\nimport mock \"github.com/stretchr/testify/mock\"\n\n" (expected)
			        != "package mocks\n\nimport mock \"github.com/stretchr/testify/mock\"\nimport test \"github.com/kevinburke/mockery/mockery/fixtures\"\n\n" (actual)
	Messages:	The generator produced the expected prologue.

--- FAIL: TestGeneratorPrologueWithImports (1.17s)
        Error Trace:    generator_test.go:64
			generator_test.go:205
	Error:		Not equal: "package mocks\n\nimport http \"net/http\"\nimport mock \"github.com/stretchr/testify/mock\"\n\n" (expected)
			        != "package mocks\n\nimport http \"net/http\"\nimport mock \"github.com/stretchr/testify/mock\"\nimport test \"github.com/kevinburke/mockery/mockery/fixtures\"\n\n" (actual)
	Messages:	The generator produced the expected prologue.

--- FAIL: TestGeneratorPrologueWithMultipleImportsSameName (1.35s)
        Error Trace:    generator_test.go:64
			generator_test.go:218
	Error:		Not equal: "package mocks\n\nimport fixtureshttp \"github.com/kevinburke/mockery/mockery/fixtures/http\"\nimport http \"net/http\"\nimport mock \"github.com/stretchr/testify/mock\"\n\n" (expected)
			        != "package mocks\n\nimport fixtureshttp \"github.com/kevinburke/mockery/mockery/fixtures/http\"\nimport http \"net/http\"\nimport mock \"github.com/stretchr/testify/mock\"\nimport test \"github.com/kevinburke/mockery/mockery/fixtures\"\n\n" (actual)
	Messages:	The generator produced the expected prologue.

--- FAIL: TestPrologueWithImportFromNestedInterface (1.51s)
        Error Trace:    generator_test.go:64
			generator_test.go:990
	Error:		Not equal: "package mocks\n\nimport fixtureshttp \"github.com/kevinburke/mockery/mockery/fixtures/http\"\nimport http \"net/http\"\nimport mock \"github.com/stretchr/testify/mock\"\n\n" (expected)
			        != "package mocks\n\nimport fixtureshttp \"github.com/kevinburke/mockery/mockery/fixtures/http\"\nimport http \"net/http\"\nimport mock \"github.com/stretchr/testify/mock\"\nimport test \"github.com/kevinburke/mockery/mockery/fixtures\"\n\n" (actual)
	Messages:	The generator produced the expected prologue.

--- FAIL: TestGeneratorSuite (40.41s)
FAIL
FAIL	github.com/kevinburke/mockery/mockery	135.937s
```

any idea what's going on? maybe renaming the imports changed the order in the file? though it looks like sometimes it's adding entirely new imports.

**Comment by kevinburke** [comment]
I also add some packages in `lib` - `$GOPATH/src/lib/rpx`. When I run `mockery --output=src/lib/testrpx --name='.*Client' --dir=./src/lib/rpx`, the generated import paths have my full file tree included:

```
import rpx "/Users/kevin/path/to/mygopath/src/lib/rpx/"
```

I'd hope for output like this:

```
import rpx "lib/rpx"
```

**Comment by evanphx** [comment]
Sorry for the delay.

Oh my goodness. This basically rewrites the whole tool. It's going to take me some time to go through it all.

**Comment by IvanMalison** [comment]
@kevinburke I think that the test failures likely have to do with the way that you've cloned the repository

See how your import paths have kevinburker e.g.: `github.com/kevinburke/mockery/mockery/fixtures`

in them.

It is an unfortunate feature of golang that you basically need to put any package in the appropriate directory in order for it to function properly. I think that you will find that if you move the package to `$GOPATH/src/github.com/vektra/mockery`, things will work as expected.

**Comment by IvanMalison** [comment]
> I also add some packages in lib - $GOPATH/src/lib/rpx. When I run mockery --output=src/lib/testrpx --name='.*Client' --dir=./src/lib/rpx, the generated import paths have my full file tree included: `import rpx "/Users/kevin/path/to/mygopath/src/lib/rpx/"`

This is pretty strange. I'm not 100% sure what is going on here. Are you sure that you properly exported GOPATH so that it is accesible to mockery? Did you merge your own changes into mine? could there be an issue there.

Let me see if I am able to reproduce.

**Comment by sidazhang** [comment]
@evanphx For what it is worth, I have been using this branch on a few teams at work for a while. It seems to work much more consistently than the master branch.

I want to make a proposal at work to drive for broader adoption of mockery (over mockgen). But the issues with imports and other bugs fixed by this pull request is a dealbreak at the moment. 

Do you have an ETA in mind for looking at this diff more? I would love to get rid of mockgen once and for all.

**Comment by IvanMalison** [comment]
I'm pretty sure that this also fixes #96

**Comment by alfred-landrum** [comment]
I wanted to add confirmation that this branch does address #96 for a repo I work on.

**Comment by sidazhang** [comment]
@kevinburke hey would you be able to indicate whether you have plans to look at this?

As a result of the various bugs in mockery, we have been forced to get everyone to install this pull request at work.

**Comment by evanphx** [comment]
Sorry for the huge delay on this! I've gone ahead and merged it.

**Pull request merged**